### PR TITLE
Introduce Sloppy

### LINKED ISSUE #2763: Consider chunking over the compressed stream of value byte serialization
When we moved to chunking over the value byte-serialization for all chunks, we observed that our realized average chunk size can vary quite a bit from the desired size because compression occurs at a layer above chunking. Thus for blob types which are already compressed (e.g. jpg), we observed average chunk size close to the target (4k), but for collections of uncompressed data (e.g. a CSV imported to a List or Map), we tend to see average sizes closer to 2k. This obviously has implications for read performance as quite a bit more chunks would be well below the disk block read size.
#2762 Raises another aspect of this issue, which is that for keyed collections (Map, Set), very long key values will essentially result in a dramatic reduction in the number of values stored in each chunk and thus a much _narrower_ prolly tree.

I ran an experiment of constructing a large map of `String(fmt.Sprintf("%d", i))` = > Number(i) and compared that to prefixing the string with a large repeated substring (similar to a sequence of long urls which vary only in the last few bytes).

As expected, the average chunk size gets roughly cut in half for every 150-200 bytes of string length.

Separately I tested compressing a file of thousands of repeats of a 200 character URL with both zlib and snappy. Surprisingly, snappy fairs well (it compresses to a factor of 10 of zlib, but it does so proportionally).

I'm curious how common b-tree sql engines handle this. I know that leveldb run-length encodes keys and forces a re-start at the block-read-size-boundary.

Chunking over the compressed stream of value bytes is an appealing potential solution because snappy appears to have a sufficiently long "memory" to handle this well enough and its a general solution that would address this problem and also get us back to a more uniform observed average chunk size.

This requires investigation along two directions:

-Understanding snappy enough to know if its safe to chunk over the current output of the compressed stream (i.e. whether it's maintaining an internal buffer of input bytes which _would_ be flushed on close).

-We'd need to re-start the compression context after each boundary, and thus there may be implications for re-chunking correctly.

### LINKED ISSUE #3369: Pass all encoded bytes to rolling value hasher
In order to compensate for non-compressed bytes in the rolling hash, we presently don't pass Ref types. This is a big hack. It was done in https://github.com/attic-labs/noms/commit/fbfdd317fc014cd2a6cc9973f45dd90bf316291a and should be removed if/when #2763 lands

### PR DESCRIPTION
**Pull request by rafael-atticlabs**
Introduce Sloppy - an estimating compression function for snappy - which allows for the rolling hash to better produce a given target chunk size *after compression*.

Fixes #2763, Fixes #3369

### CODE REVIEW

**Comment by rafael-atticlabs** [comment]
FYI, tests of encoding structured data show that this reduces the number of chunks by about 35-40% and the overall size on disk by about 10-15%.

File: go/types/value_encoder.go [hunk 1]
```diff
@@ -13,11 +13,10 @@ import (
 
 type valueEncoder struct {
 	nomsWriter
-	forRollingHash bool
 }
 
-func newValueEncoder(w nomsWriter, forRollingHash bool) *valueEncoder {
-	return &valueEncoder{w, forRollingHash}
+func newValueEncoder(w nomsWriter) *valueEncoder {
```
Comment (arv) [inline go/types/value_encoder.go]: Yes! This will make my code cleaner too

File: go/sloppy/sloppy_test.go [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+		// Prefer closer match
+		{128, "23hello my friend, 12hello my friend, 01hello my friend, 23hello my friendxxxx",
+			[]interface{}{"23hello my friend, 12", uint16(19), uint16(17),
+				"01", uint16(19), uint16(17),
+				"23", uint16(19), uint16(15),
+				"x",
+			}},
+	}
+
+	for _, c := range tc {
```
Comment (arv) [inline go/sloppy/sloppy_test.go]: Use local test function instead. Table driven tests don't tell you which one fails.

File: go/sloppy/sloppy.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package sloppy
+
+import "github.com/attic-labs/noms/go/d"
+
+// Sloppy is a logical variant of Snappy. Its purpose to provide a kind of
```
[the remark this answers is not in the document]
Comment (rafael-atticlabs) [reply go/sloppy/sloppy.go]: done

File: go/sloppy/sloppy.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+// estimate of how a given byte sequence *will be* compressed by Snappy. It is
+// useful when a byte stream is fed into a rolling hash with the goal of
+// achieving a given average chunk byte length *after compression*. Sloppy is
+// logically similar to snappy, but prefers "copies" which are closer to the
+// repeated byte sequence (snappy prefers to refer to the *first* instance of a
+// repeated byte sequence).
```
[the remark this answers is not in the document]
Comment (rafael-atticlabs) [reply go/sloppy/sloppy.go]: done

File: go/sloppy/sloppy.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+// bytes earlier in the stream. Snappy and sloppy behave roughly identical thus
+// far.
+//
+// When in the "matched state" (attempting to extend the current match), Snappy
+// does not re-index new 4-byte sequences, but Sloppy does. The reason for this
+// is that Sloppy would like match the most recent occurence as it moves
+// forward.
+//
+// Lastly, sloppy implements two novel heuristics, both aimed at mitigating
+// the liklihood that altering any byte in an input stream will cause chunk
```
[the remark this answers is not in the document]
Comment (rafael-atticlabs) [reply go/sloppy/sloppy.go]: done

File: go/sloppy/sloppy.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	for ; from < to; from++ {
+		if sl.enc.emitLiteral(src[from]) {
+			return true
+		}
+	}
+	return false
+}
+
+// Emit a copy if the length is sufficient for a given offset
+func (sl *Sloppy) maybeCopy(src []byte, i int) bool {
```
[the remark this answers is not in the document]
Comment (rafael-atticlabs) [reply go/sloppy/sloppy.go]: done

File: go/sloppy/sloppy.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+	// 8 LSBits of length
+	if be.f(byte(length)) {
+		return true
+	}
+
+	return false
+}
+
+func fbhash(u, shift uint32) uint32 {
+	return (u * 0x1e35a7bd) >> shift
```
[the remark this answers is not in the document]
Comment (rafael-atticlabs) [reply go/sloppy/sloppy.go]: done

File: go/types/rolling_value_hasher.go [hunk 8]
```diff
@@ -64,86 +66,39 @@ func hashValueByte(item sequenceItem, rv *rollingValueHasher) {
...
+		bw:      bw,
+		enc:     enc,
 		bz:      buzhash.NewBuzHash(window),
 		pattern: pattern,
 		window:  window,
 		salt:    salt,
 	}
-	rv.enc = newValueEncoder(rv, true)
+
+	rv.sl = sloppy.New(func(b byte) bool { return rv.HashByte(b) })
```
[the remark this answers is not in the document]
Comment (rafael-atticlabs) [reply go/types/rolling_value_hasher.go]: done

File: go/types/value_encoder.go (hunk 1 above)
Comment (rafael-atticlabs) [reply go/types/value_encoder.go]: :-)

File: go/sloppy/sloppy_test.go (hunk 2 above)
Comment (rafael-atticlabs) [reply go/sloppy/sloppy_test.go]: done

**Pull request merged**